Scout parsing improvements - phase 2

bfo1017 opened the pull request.

timmyoder committed:
scout raw parsing stub

bfo1017 committed:
Map Scout measures to corresponding technology classes in bicep.technologies table

bfo1017 committed:
Update .gitignore to exclude local dev and troubleshooting files

bfo1017 committed:
Restore adoption_forecast_parsing.py to original structure after EDA

bfo1017 committed:
Add initial version of technology_stock_parser.py for data processing

bfo1017 committed:
Add local file processing for Scout, EV, and PV technology adoption forecasts

bfo1017 committed:
Add local state cost factors processing function

bfo1017 committed:
Implement local data processing and configurable state filtering for user friendly BICEP

bfo1017 committed:
Make CapacityEstimate configurable for any state(s)

bfo1017 committed:
Implement local cost factors and configurable state support

bfo1017 committed:
Make BicepResults configurable with required target_states parameter

bfo1017 committed:
Add local input data files for user-based deployment

bfo1017 committed:
Update gitignore for local data processing files

timmyoder committed:
add config with dir parsing

bfo1017 committed:
Consolidate config systems and implement parsed input  table saving

bfo1017 committed:
fix bug for saving DB and local mode

timmyoder committed:
don't track any data files

Data files aren't directly tracked in git because they are too large. Any required data files will be added to releases as assets available for download

timmyoder committed:
script to save ms sql db tables to sqlite local file

timmyoder committed:
allow query from cloud db or local sqlite file

timmyoder committed:
TODOs for checking and downloading github data file assets

timmyoder committed:
TODO for target_states parameter

bfo1017 committed:
Add save functionality with state suffix

bfo1017 committed:
Fix state filtering in capacity calculations

bfo1017 committed:
Add target states filtering and caching

bfo1017 committed:
Add target states validation and mode support

bfo1017 committed:
Add local file support for cost factors and PV data

bfo1017 committed:
Add data path configuration and asset validation

bfo1017 committed:
Clean up adoption forecast parsing

bfo1017 committed:
Update database upload utilities

bfo1017 committed:
Add local database mirroring utilities

bfo1017 committed:
Remove unused technology stock parser

bfo1017 committed:
edited full results saving as requested by Tim

bfo1017 committed:
Fix base year handling

bfo1017 committed:
Add individual state analysis following Tim's approach

bfo1017 committed:
Integrate all_states method and verify sync

bfo1017 committed:
Delete bicep/all_states.py

integrated into analysis

timmyoder committed:
add parquet option

bfo1017 committed:
Remove unused file

bfo1017 committed:
Merge branch 'scout-parsing' of github.com:pnnl/BICEP into scout-parsing

timmyoder commented on utils/local_db_mirror.py:
Do we need this module still?

bfo1017 replied:
Yes, we can remove this module. It's not imported anywhere and was replaced by the direct parquet file handling in data assets.

timmyoder replied:
The file is still showing up so please remove :)

bfo1017 replied:
removed it, Tim! it shouldn't show up anymore

timmyoder commented on utils/db_models.py:
```diff
@@ -210,33 +230,39 @@ def get_state_cost_factors():
-    FROM dbo.CountyHierarchy
+    pandas.DataFrame: DataFrame with columns ['State', 'Factor']
     """
+    import os
```
if we need this import, move it to the top of the file with the rest of the imports. if we don't need, removeit.

bfo1017 replied:
Removed the import. It wasn't being used

timmyoder commented on utils/db_models.py:
```diff
@@ -210,33 +230,39 @@ def get_state_cost_factors():
 
 
-def get_new_pv_data():
+def get_state_cost_factors_local():
```
The cost factor data is included in the database, correct? if so, we shouldn't be loading it in from the csv, but pull it in from the database

timmyoder commented on utils/db_models.py:
```diff
@@ -8,7 +8,8 @@
 import pandas as pd
-
+from pathlib import Path
+from utils.config import COST_FACTOR_FILE
```
if we no longer need the csv file for the cost factors (contained in the database), lets remove this import. if it is still needed, move to be with the other imports from utils.config (line 23 currently).

bfo1017 replied:
Fixed, now loading from database using StateCostFactors table instead of CSV file.

timmyoder commented on data/required_input/cost_factor.csv:
This data is in the database, so we don't need this file i believe.

bfo1017 replied:
Fixed

timmyoder commented on utils/adoption_forecast_parsing.py:
```diff
@@ -4,23 +4,98 @@
+        return None
+    return str(scenario_name).strip().lower()
+
+def standardize_dataframe_for_bicep_v1(df):
```
why _v1? is there a v2? let's rename to remove that from the name?

bfo1017 replied:
Removed the _v1 suffix. There's only one version now.

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+
+# Root directory of the BICEP project (utils/../ = project root)
+BICEP_ROOT = Path(__file__).parent.parent
+ROOT = BICEP_ROOT  # Keep Tim's variable name for compatibility
```
Remove one of these: [BICEP_ROOT, ROOT]. Looks like you don't use the ROOT anywhere, so i suggest removing that one.

bfo1017 replied:
Removed ROOT, kept BICEP_ROOT

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+
+# Data directories
+DATA_ROOT = BICEP_ROOT / 'data'
+DATA_DIR = DATA_ROOT  # Keep Tim's variable name for compatibility
```
Same comment here about removing the duplicate name.

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+DATA_DIR = DATA_ROOT  # Keep Tim's variable name for compatibility
+RAW_INPUTS_PATH = DATA_ROOT / 'raw_inputs'
+PARSED_INPUTS_PATH = DATA_ROOT / 'parsed_inputs'
+REQUIRED_INPUT_PATH = DATA_ROOT / 'required_input'
```
Remove

bfo1017 replied:
Removed DATA_DIR, kept DATA_ROOT

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+
+# Hierarchy and mapping files
+HIERARCHY_FILE = RAW_INPUTS_PATH / 'hierarchy.csv'
+TECHNOLOGY_MAP_FILE = BICEP_ROOT / 'technology_map.csv'
```
whats this file? do we use it at all?

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+    logger.error(error)
+    raise AssertionError
+
+# ============= INPUT FILES =============
```
```suggestion
# ============= TECHNOLOGY PROJECTION INPUT FILES =============
```

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+    """
+    import urllib.request
+    import time
+    from pathlib import Path
```
Import should go at the top of the file.

timmyoder replied:
Still outstanding

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+                     'https://github.com/pnnl/BICEP/releases/download/v0.1-data/stock-meta.parquet',
+                     'https://github.com/pnnl/BICEP/releases/download/v0.1-data/technologies.parquet',
+                     'https://github.com/pnnl/BICEP/releases/download/v0.1-data/upgrades.parquet',
+                     'https://github.com/pnnl/BICEP/releases/download/v0.1-data/bicep.x-stock.db'
```
missing https://github.com/pnnl/BICEP/releases/download/v0.1-data/CountyHierarchy.parquet and https://github.com/pnnl/BICEP/releases/download/v0.1-data/state_cost_factors.parquet

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+    logger.error(f'DATA_LOCATION {DATA_LOCATION} is not valid. Must be in ["LOCAL", "PNNL Database"]')
+    logger.error(error)
+    raise AssertionError
+
```
Let's add logic here to run check for the local files if it is in "LOCAL" model. 

```suggestion
if DATA_LOCATION == 'LOCAL':
    ensure_data_assets()
```

timmyoder commented on utils/config.py:
```diff
@@ -0,0 +1,251 @@
+
+
+def ensure_data_assets():
+    """
```
What you are doing right now is checking if each file exists, and if one of them doesn't exist, you download all the files. A more efficient way to handle this would be to loop through the expected files, and if it doesn't exist, just download only that file. That way you'd only download the files you don't have and not the ones that are already downloaded. This works though so don't worry about updating now. just for future reference.

timmyoder replied:
Still outstanding

timmyoder commented on bicep/upgrades.py:
```diff
@@ -111,8 +141,8 @@ def _upgrade_costs(self):
-            logger.info('Retrieving state location factors from database')
-            state_factors = get_state_cost_factors()
+            logger.info('Retrieving state location factors from local file')
+            state_factors = get_state_cost_factors_local()
```
See comment about reading this data from the database (either local or pnnl cloud) instead of the csv.

timmyoder commented on bicep/capacity.py:
```diff
@@ -125,7 +130,7 @@ class CapacityEstimate:
     def __init__(self, residential_voltage=240, commercial_voltage=480,
                  medium_voltage=12470, max_light_comm_amp=1000, ev_charger_amp=50,
-                 panel_safety_factor=1.25):
+                 panel_safety_factor=1.25, target_states=None):
```
We should pass the self.mode parameter to this class as well. The building stock metadata and peak load data should be pulled from the same location (local or pnnl database) as the other data (tech adoption, etc).

timmyoder commented on bicep/capacity.py:
```diff
@@ -49,12 +49,17 @@
                                        PeakLoad.residential == residential)
     else:
         query = select(PeakLoad).where(PeakLoad.upgrade == upgrade)
+    
```
Make sure this accommodates the data location mode parameter.

timmyoder commented on bicep/capacity.py:
```diff
@@ -147,16 +156,18 @@ def __init__(self, residential_voltage=240, commercial_voltage=480,
     def get_meta(self):
         logger.info('Getting stock metadata')
-        self.building_meta = query_to_df(select(StockMeta).where(StockMeta.state == 'CA'))
+        self.building_meta = query_to_df(select(StockMeta).where(StockMeta.state.in_(self.target_states)))
```
Make sure this accommodates the data location mode parameter.

timmyoder commented on bicep/capacity.py:
```diff
@@ -231,8 +242,8 @@ def building_req_capacity(self, hp_upgrades=(3, 4,), hpwh_upgrades=(6, )):
-        # retrieve calculated load differences
-        all_upgrades = query_to_df(select(LoadDifference))
+        # retrieve calculated load differences - filter by target states
+        all_upgrades = query_to_df(select(LoadDifference).where(LoadDifference.state.in_(self.target_states)))
```
Make sure this accommodates the data location mode parameter.

timmyoder committed:
Refactor: Replace global DATA_LOCATION with DatabaseContext dependency injection

- Add DatabaseContext class for clean database configuration management
- Update all classes to use dependency injection instead of global state

timmyoder committed:
Ensure local data exists if db context is "local"

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +218,139 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+        PARSED_INPUTS_PATH.mkdir(parents=True, exist_ok=True)
+        
+        # Save aggregate cost summary results
+        self.all_states_results.to_csv(f"data/parsed_inputs/bicep_cost_summary_{self.scenario}.csv", index=False)
```
You should use the pathlib path object PARSED_INPUTS_PATH instead of hard coding the directory here.

```suggestion
        self.all_states_results.to_csv(PARSED_INPUTS_PATH / f"bicep_cost_summary_{self.scenario}.csv", index=False)
```

timmyoder replied:
You should use the pathlib path when saving instead of hard coding that path. 

```suggestion
        self.all_states_results.to_csv(PARSED_INPUTS_PATH / f"bicep_cost_summary_{self.scenario}.csv", index=False)
```

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +218,139 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+        self.all_states_meta.to_csv(f"data/parsed_inputs/bicep_meta_{self.scenario}_all_states.csv", index=False)
+        
+        logger.info(f"Saved results to data/parsed_inputs/")
+
```
Change all these hard coded paths to the Pathlib version like the comment above

timmyoder commented on bicep/analysis.py:
```diff
@@ -28,12 +34,103 @@ def __init__(self, aggregation_level='state', annualized=True,
+        if isinstance(self.target_states, list):
+            if len(self.target_states) == 1:
+                state = self.target_states[0]
+            else:
```
what about if someone wants to analyze more than one state, but not all? (e.g., ['WA', "CA", "PA"]. i think the way this is coded just gives the option for one state or all states and nothing in between.

timmyoder commented on bicep/analysis.py:
```diff
@@ -28,12 +34,103 @@ def __init__(self, aggregation_level='state', annualized=True,
+        elif isinstance(self.target_states, str):
+            state = self.target_states
+        else:
+            state = self.buildings['state'].iloc[0]
```
why do you have this one here? this could give the user unexpected results.

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +218,139 @@ def plot_panel_capacity(self, residential=1, log_y=True):
-if __name__ == '__main__':
-    bau = BicepResults(scenario='bau')
-    high = BicepResults(scenario='high')
+class BicepAllStates:
```
We should modify this so it is capable of running any set of states (just one, a subset of the states, or all states) instead of just all or nothing.

timmyoder replied:
Since this can run any subset of states and not just "All", we should rename this to something like BicepMultiStateResults or something similar

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +218,139 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+        # Save meta data
+        self.all_states_meta.to_csv(f"data/parsed_inputs/bicep_meta_{self.scenario}_all_states.csv", index=False)
+        
+        logger.info(f"Saved results to data/parsed_inputs/")
```
change all of these to_csv calls to use the pathlib object

timmyoder commented on bicep/upgrades.py:
```diff
@@ -36,14 +39,48 @@ def __init__(self, aggregation_level='state', annualized_costs=True, upgrade_lif
+
+        # Ensure data assets are available before proceeding
+        from utils.config import ensure_data_assets
+        ensure_data_assets()
```
let's remove this import and data asset check. we should only be doing that when it's running in local mode and it is already happening in the db_models model

timmyoder commented on bicep/upgrades.py:
```diff
@@ -36,14 +39,48 @@ def __init__(self, aggregation_level='state', annualized_costs=True, upgrade_lif
+                raise ValueError(f"Invalid state(s): {invalid_states}. Valid states are: {valid_states}")
+            logger.info(f"Analyzing target states: {target_states}")
+        else:
+            raise ValueError("target_states must be 'all', a state abbreviation (e.g., 'CA'), or a list of state abbreviations (e.g., ['CA', 'TX'])")
```
This is great! we should use this same logic in the analysis module. see my comments in that file.

timmyoder commented on bicep/tech_adoption.py:
```diff
@@ -8,14 +8,17 @@
+import sys
+
+logger.remove()
+logger.add(sys.stderr, level="INFO")
```
I don't think we need to remove and add the logger again. let's just keep the default one we import.

timmyoder commented on bicep/tech_adoption.py:
```diff
@@ -54,87 +71,116 @@ def calculate_adoptions(self):
             raise KeyError(f"Technology must be in {self.all_techs['tech_name'].to_list()}")
 
+        # Load unified adoption forecasts data (contains all scenarios)
+        cache_key = f"unified_{self.mode}"
```
what is the purpose of the cache here? it looks like you are reading all the tech projects for all states, sectors, and techs. i think it might be better to just query the data we want to pull from the database and just pull that filtered data instead of selecting all

timmyoder reviewed (changes requested):
This is looking great Bilal. Most comments are minor. Biggest change requested is to make the bicep.analysis classes capable of running an arbitrary set of states (one, a subset of all states, and all the states).

i also made a change to how we were handing the data location parameter.

bfo1017 committed:
Remove duplicate config variables

bfo1017 committed:
Fix DATA_ROOT import

bfo1017 committed:
Remove v1 suffix

bfo1017 committed:
Add db_context parameter

bfo1017 committed:
Pass db_context correctly

bfo1017 committed:
Remove logger config, optimize queries

bfo1017 committed:
Add pathlib, state validation

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +220,162 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+    
+    def _run_individual_state_analysis(self):
+        """Run individual state analysis with state-by-state processing"""
+        import pandas as pd
```
Move this import to the top of the file please.

bfo1017 replied:
Moved import pandas as pd to the top of bicep/analysis.py with the other imports.

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +220,162 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+    
+    def _save_individual_state_results(self):
+        """Save results from individual state analysis with standard naming convention"""
+        from utils.config import PARSED_INPUTS_PATH
```
Move this import to the top of the file please.

bfo1017 replied:
Moved from utils.config import PARSED_INPUTS_PATH to the top of bicep/analysis.py

timmyoder commented on bicep/analysis.py:
```diff
@@ -28,12 +34,105 @@ def __init__(self, aggregation_level='state', annualized=True,
+    
+    def _create_cost_summary(self):
+        """Create cost summary by state and building type."""
+        import pandas as pd
```
Move this import to top of file please.

bfo1017 replied:
The import is now at the top

timmyoder commented on bicep/analysis.py:
```diff
@@ -121,16 +220,162 @@ def plot_panel_capacity(self, residential=1, log_y=True):
+        
+        # Save results if requested
+        if self.save_results:
+            self._save_individual_state_results()
```
If i'm tracking this correctly, this will save the full results for all states and not just individual states right? Like we aren't creating 50 separate files for each state, it is saving a summary file, buildings data from all states, and metadata from all states, right? perhaps we should rename the _save_individual_state_results() function to something else like save_results().

bfo1017 replied:
You're correct! Saves combined buildings + metadata across all states. I Renamed _save_individual_state_results() to save_results() as advised Tim!

timmyoder commented on bicep/analysis.py:
```diff
@@ -28,12 +34,105 @@ def __init__(self, aggregation_level='state', annualized=True,
+    def _save_results(self):
+        """Save results to CSV with state abbreviation."""
+        from utils.config import PARSED_INPUTS_PATH
+        import pandas as pd
```
move import to top of file.

timmyoder commented on utils/adoption_forecast_parsing.py:
```diff
@@ -4,23 +4,98 @@
 from io import BytesIO
 import json
+from pathlib import Path
+import logging
```
Let's be consistent with the other files and use loguru here instead of the logging library.

```suggestion
from loguru import logger
```

timmyoder commented on utils/adoption_forecast_parsing.py:
```diff
@@ -4,23 +4,98 @@
 from utils.sensitive_config import AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY
+from utils.config import SCOUT_BAU_FILE, SCOUT_HIGH_FILE, EV_PROJECTIONS_FILE, HIERARCHY_FILE, PV_PROJECTIONS_FILE
 
+logger = logging.getLogger(__name__)
```
we can remove this if we are using loguru. see comment above

timmyoder reviewed (changes requested):
Looking pretty good Bilal! just a few more changes and a few outstanding comments that you hadn't quite completed. Almost there!

bfo1017 committed:
Move imports to top, rename class and save method

bfo1017 committed:
Switch from logging to loguru

bfo1017 committed:
Move inline imports to top of file

bfo1017 committed:
Remove legacy local_db_mirror.py

bfo1017 committed:
Fix naming conflict and FutureWarning in concat

bfo1017 committed:
Fix FutureWarning in fillna downcasting

bfo1017 committed:
Add documentation for running modes and state selection

bfo1017 committed:
Add visualization and output data documentation

bfo1017 committed:
Remove docs (moving to docs branch)

bfo1017 committed:
Move inline imports to top of config.py

bfo1017 commented:
I went through your PR comments and fixed the ones I missed, Tim, thanks for catching those. I tested BICEP in Azure + local mode and results match

timmyoder reviewed (approved):
Looks great! Thanks for making all those changes.
